File: 2015/7/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type InstructionType int

const (
	AND InstructionType = iota
	OR
	LSHIFT
	RSHIFT
	NOT
	ASSIGN
)

type Instruction struct {
	operation InstructionType
	input1    string
	input2    string
	output    string
}

var instructions []Instruction
var hashmap map[string]int

func main() {
	args := os.Args[1:]

	if len(args) < 1 {
		fmt.Println("Usage: go run main.go <input.txt>")
		return
	}

	var filename string = args[0]

	file, err := os.Open(filename)

	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}

	defer file.Close()

	instructions = make([]Instruction, 0)
	hashmap = make(map[string]int)

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		InsertInstruction(line)
	}

	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	// Part 1
	a_val := Evaluate("a")
	fmt.Println("Part 1:", a_val)

	// Part 2
	hashmap = make(map[string]int)
	hashmap["b"] = a_val
	a_val = Evaluate("a")
	fmt.Println("Part 2:", a_val)
}

func InsertInstruction(line string) {
	var instruction Instruction

	if strings.Contains(line, "AND") {
		instruction.operation = AND
		fmt.Sscanf(line, "%s AND %s -> %s", &instruction.input1, &instruction.input2, &instruction.output)
	} else if strings.Contains(line, "OR") {
		instruction.operation = OR
		fmt.Sscanf(line, "%s OR %s -> %s", &instruction.input1, &instruction.input2, &instruction.output)
	} else if strings.Contains(line, "LSHIFT") {
		instruction.operation = LSHIFT
		fmt.Sscanf(line, "%s LSHIFT %s -> %s", &instruction.input1, &instruction.input2, &instruction.output)
	} else if strings.Contains(line, "RSHIFT") {
		instruction.operation = RSHIFT
		fmt.Sscanf(line, "%s RSHIFT %s -> %s", &instruction.input1, &instruction.input2, &instruction.output)
	} else if strings.Contains(line, "NOT") {
		instruction.operation = NOT
		fmt.Sscanf(line, "NOT %s -> %s", &instruction.input1, &instruction.output)
	} else {
		instruction.operation = ASSIGN
		fmt.Sscanf(line, "%s -> %s", &instruction.input1, &instruction.output)
	}

	instructions = append(instructions, instruction)
}

func Evaluate(find_target string) int {
	// check if the value is already in the hashmap
	if val, ok := hashmap[find_target]; ok {
		return val
	}

	var instruction Instruction

	//find the instruction that has the target as output
	for _, inst := range instructions {
		if inst.output == find_target {
			instruction = inst
			break
		}
	}

	var input1_val, input2_val int = 0, 0

	if instruction.input1 != "" {
		if _, err := fmt.Sscanf(instruction.input1, "%d", &input1_val); err != nil {
			input1_val = Evaluate(instruction.input1)
		}
	}

	if instruction.input2 != "" {
		if _, err := fmt.Sscanf(instruction.input2, "%d", &input2_val); err != nil {
			input2_val = Evaluate(instruction.input2)
		}
	}

	switch instruction.operation {
	case InstructionType(AND):
		hashmap[find_target] = input1_val & input2_val
		break
	case InstructionType(OR):
		hashmap[find_target] = input1_val | input2_val
		break
	case InstructionType(LSHIFT):
		hashmap[find_target] = input1_val << input2_val
		break
	case InstructionType(RSHIFT):
		hashmap[find_target] = input1_val >> input2_val
		break
	case InstructionType(NOT):
		hashmap[find_target] = ^input1_val
		break
	case InstructionType(ASSIGN):
		hashmap[find_target] = input1_val
		break
	}

	return hashmap[find_target]
}

File: 2015/lua/7/main.lua
InstructionType = {
    AND = "AND",
    OR = "OR",
    LSHIFT = "LSHIFT",
    RSHIFT = "RSHIFT",
    NOT = "NOT",
    ASSIGN = "ASSIGN"
}

Instructions = {}

Wires = {}

function get_signal(target)
    if tonumber(target) then
        return tonumber(target)
    end

    if Wires[target] then
        return Wires[target]
    end

    local instruction = nil
    for i, inst in ipairs(Instructions) do
        if inst.output == target then
            instruction = inst
            break
        end
    end

    if not instruction then
        return nil
    end

    local input1 = tonumber(instruction.input1) or get_signal(instruction.input1)
    local input2 = nil
    if instruction.input2 then
        input2 = tonumber(instruction.input2) or get_signal(instruction.input2)
    end

    local output = 0
    if instruction.type == InstructionType.AND then
        output = input1 & input2
    elseif instruction.type == InstructionType.OR then
        output = input1 | input2
    elseif instruction.type == InstructionType.LSHIFT then
        output = input1 << input2
    elseif instruction.type == InstructionType.RSHIFT then
        output = input1 >> input2
    elseif instruction.type == InstructionType.NOT then
        output = ~input1
    elseif instruction.type == InstructionType.ASSIGN then
        output = input1
    end

    Wires[target] = output
    return output
end

local function main()
    if #arg < 1 then
        print("Usage: lua main.lua <input.txt>")
        os.exit(1)
    end

    local filename = arg[1]
    local file = io.open(filename, "r")
    if not file then
        print("Error: Could not open file " .. filename)
        os.exit(1)
    end

    local lines = {}
    for line in file:lines() do
        table.insert(lines, line)
    end

    file:close()

    for i, line in ipairs(lines) do
        local instruction = nil
        --split line by space
        local parts = {}
        for part in string.gmatch(line, "%S+") do
            table.insert(parts, part)
        end

        if #parts == 3 then
            instruction = {
                type = InstructionType.ASSIGN,
                input1 = parts[1],
                output = parts[3]
            }
        elseif #parts == 4 then
            instruction = {
                type = InstructionType.NOT,
                input1 = parts[2],
                output = parts[4]
            }
        elseif #parts == 5 then
            if parts[2] == "AND" then
                instruction = {
                    type = InstructionType.AND,
                    input1 = parts[1],
                    input2 = parts[3],
                    output = parts[5]
                }
            elseif parts[2] == "OR" then
                instruction = {
                    type = InstructionType.OR,
                    input1 = parts[1],
                    input2 = parts[3],
                    output = parts[5]
                }
            elseif parts[2] == "LSHIFT" then
                instruction = {
                    type = InstructionType.LSHIFT,
                    input1 = parts[1],
                    input2 = parts[3],
                    output = parts[5]
                }
            elseif parts[2] == "RSHIFT" then
                instruction = {
                    type = InstructionType.RSHIFT,
                    input1 = parts[1],
                    input2 = parts[3],
                    output = parts[5]
                }
            end
        end

        table.insert(Instructions, instruction)
    end

    local a = get_signal("a")
    print("Part 1: " .. a)

    Wires = {}
    Wires["b"] = a
    a = get_signal("a")
    print("Part 2: " .. a)
end

main()
